gen/json: Enable omitempty for optional collections

ravirajj opened the pull request:
This changes code generation to enable the `omitempty` JSON option for
optional collections to reduce `null` noise in the JSON output.

So,

    struct OptionalListField {
        1: optional list<Item> items
        2: optional map<Key, Value> dictionary
    }

Becomes,

    type OptionalListField struct {
      Items       []Item         `json:"items,omitempty"`
      Dictionary  map[Key]Value  `json:"dictionary,omitempty"`
    }

Encoding `OptionalListField{}` will now produce, `{}` instead of,

    {"items": null, "dictionary": null}

Resolves #330

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/thriftrw/thriftrw-go?pullRequest=331) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/thriftrw/thriftrw-go?pullRequest=331) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/thriftrw/thriftrw-go?pullRequest=331) it.</sub>

kriskowal commented on compile/container.go:
```diff
@@ -93,14 +93,14 @@ func (m *MapSpec) TypeCode() wire.Type {
 // ForEachTypeReference for MapSpec
-func (m *MapSpec) ForEachTypeReference(f func(TypeSpec) error) error {
-	if err := f(m.KeySpec); err != nil {
+func (m *MapSpec) ForEachTypeReference(f func(TypeSpec) error) (err error) {
```
The named return is inconsistent with the style below. We rarely use named returns, but when you do, the `return` statement receives no arguments. This also appears to be a style change unrelated to the PR.

ravirajj replied:
This change was made to fix the lint error.  I'll update the style.

kriskowal commented on gen/field.go:
```diff
@@ -147,23 +147,24 @@ func generateTags(f *compile.FieldSpec) (string, error) {
 	}
 
-	if (isStructType(f.Type) || isPrimitiveType(f.Type)) && !f.Required && !t.HasOption("omitempty") {
+	// We want to add the "omitempty" json tag if the field is an 'optional" to
```
Nit.

Either:

    is optional

Or

    is an "optional".

Also, mismatch single and double quote.

ravirajj replied:
Done.

kriskowal commented on gen/field.go:
```diff
@@ -147,23 +147,24 @@ func generateTags(f *compile.FieldSpec) (string, error) {
 
-	if (isStructType(f.Type) || isPrimitiveType(f.Type)) && !f.Required && !t.HasOption("omitempty") {
+	// We want to add the "omitempty" json tag if the field is an 'optional" to
+	// reduce "null" noise. The "omitempty" json tag specifies that the field
```
capitalize JSON

ravirajj replied:
Done.

kriskowal commented on gen/struct_test.go:
```diff
@@ -877,112 +877,131 @@ func TestStructWithDefaults(t *testing.T) {
-		v interface{}
-		j string
+		v, w interface{}
+		j    string
```
`v` is obviously short for `value`. `j` is also obviously for `JSON`. What does `w` mean? It might be useful to extract the three cases that make use of the `w` field from the table test, stand-alone. Using positional fields for a {k, v} duple is pretty clean.

kriskowal commented on gen/struct_test.go:
```diff
@@ -992,10 +1011,13 @@ func TestStructJSON(t *testing.T) {
+		if tt.w != nil {
+			u = tt.w
+		}
+		v := reflect.New(reflect.TypeOf(u).Elem()).Interface()
```
Also might be better to use full names here, since they are drifting from being abbreviations of meaningful names. `v` in scope no longer means value, but “type of value”. `w` appears to be an alternate value for purposes of inferring the type from a more complete object.

This would probably all be a lot more clear with the three relevant cases extracted to a clearer test with meaningful names.

kriskowal commented on gen/struct_test.go:
```diff
@@ -877,112 +877,131 @@ func TestStructWithDefaults(t *testing.T) {
-		{&tu.Document{Pdf: td.PDF{}}, `{"pdf":""}`},
+		{v: &tu.EmptyUnion{}, j: "{}"},
+		{v: &tu.Document{Pdf: td.PDF("hello")}, j: `{"pdf":"aGVsbG8="}`},
+		{v: &tu.Document{Pdf: td.PDF{}}, w: &tu.Document{Pdf: nil}, j: `{}`},
```
Does the previously written test case continue to pass? Might be better to just replace, remove, or move.

abhinav commented:
Thanks for the PR! Please address @kriskowal's comments and drop styling changes that are not relevant to this PR. Also, please rebase this change on dev. I've already updated the merge base for the PR to dev.

ravirajj commented:
- Rebased on `dev`.
- Added separate tests for optional empty list/map struct fields.
- `TestStructJSON` already covers optional nil and non-empty list/map struct fields.

kriskowal commented on compile/container.go:
```diff
@@ -97,10 +97,7 @@ func (m *MapSpec) ForEachTypeReference(f func(TypeSpec) error) error {
-		return err
-	}
-	return nil
+	return f(m.ValueSpec)
```
Again with the unrelated style changes ;)

ravirajj replied:
I had to make these changes to fix the lint errors!

abhinav replied:
This is a lint error in Go 1.8 but not in 1.9. I changed Travis on this repo
to run for Go 1.9 only. This should no longer be a problem.

kriskowal commented:
Please get a stamp from @abhinav as well. We require two stamps for software networking projects.

abhinav commented:
Merged. Thanks for the change!
